crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data.rs

```diff
@@ -216,6 +220,61 @@ impl SenderData {
         Self::UnknownDevice { legacy_session: true, owner_check_failed: false }
     }
 
+    /// Create a [`SenderData`] representing the current verification state of
+    /// the given device.
+    ///
+    /// Depending on whether the device is correctly cross-signed or not, and
+    /// whether the user has been verified or not, this can return
+    /// [`SenderData::DeviceInfo`], [`SenderData::VerificationViolation`],
+    /// [`SenderData::SenderUnverified`] or [`SenderData::SenderVerified`]
+    pub fn from_device(sender_device: &Device) -> Self {
+        // Is the device cross-signed?
+        // Does the cross-signing key match that used to sign the device?
+        // And is the signature in the device valid?
+        let cross_signed = sender_device.is_cross_signed_by_owner();
+
+        if cross_signed {
+            Self::from_cross_signed_device(sender_device)
+        } else {
+            // We have device keys, but they are not signed by the sender
+            SenderData::device_info(sender_device.as_device_keys().clone())
+        }
+    }
+
+    fn from_cross_signed_device(sender_device: &Device) -> Self {
+        let user_id = sender_device.user_id().to_owned();
+        let device_id = Some(sender_device.device_id().to_owned());
+
+        let master_key = sender_device
+            .device_owner_identity
+            .as_ref()
+            .and_then(|i| i.master_key().get_first_key());
+
+        if let Some(master_key) = master_key {
+            // We have user_id and master_key for the user sending the to-device message.
+            let master_key = Box::new(master_key);
+            let known_sender_data = KnownSenderData { user_id, device_id, master_key };
+            if sender_device.is_cross_signing_trusted() {
+                Self::SenderVerified(known_sender_data)
+            } else if sender_device
+                .device_owner_identity
+                .as_ref()
+                .expect("User with master key must have identity")
```

Comment:
I know that this is existing code but this could have been expressed a bit differently to avoid the panic.

Something like:

```rust
match (device_owner, master_key) {
    (Some(..), Some(..) => {
        // Happy path
    },
    (_, _) => {
        // Surprisingly, there was no key in the MasterPubkey. We did not expect this:
        // treat it as if the device was not signed by this master key.
        //
        error!("MasterPubkey for user {user_id} does not contain any keys!");
        Self::device_info(sender_device.as_device_keys().clone())
    } 
}

Reply:
Yup. done in 8e762662b.